### PR TITLE
Add validator for repo

### PR DESCRIPTION
**Pull request by Hevienz**

### CODE REVIEW

File: pkg/manager/repo/handler.go [hunk 1]
```diff
@@ -67,17 +67,25 @@ func (p *Server) DescribeRepos(ctx context.Context, req *pb.DescribeReposRequest
 }
 
 func (p *Server) CreateRepo(ctx context.Context, req *pb.CreateRepoRequest) (*pb.CreateRepoResponse, error) {
-	// TODO: validate CreateRepoRequest
+	url := req.GetUrl().GetValue()
```
Comment (zheng1) [inline pkg/manager/repo/handler.go]: Repo needs to add a field to store the type of repo [http(s)/qingstor]

File: pkg/manager/repo/validator.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+	u, err := neturl.ParseRequestURI(url)
+	if err != nil {
+		return fmt.Errorf("Url parse failed, %s", err)
+	}
+
+	m := compRegEx.FindStringSubmatch(u.Host)
+	logger.Debugf("Regexp result: %+v", m)
+
+	if len(m) != 0 && len(m) == 3 {
```
Comment (zheng1) [inline pkg/manager/repo/validator.go]: This can be distinguished by the type of repo.

File: pkg/manager/repo/validator.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+		bucket := m[1]
+		zone := m[2]
+
+		var dat map[string]string
+		err = json.Unmarshal([]byte(credential), &dat)
+		if err != nil {
+			return fmt.Errorf("Json decode failed on credential, %+v", err)
+		}
+		access_key_id, ok := dat["access_key_id"]
```
Comment (zheng1) [inline pkg/manager/repo/validator.go]: The credential of qingstor should be a general struct.

File: pkg/manager/repo/validator.go [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+var(
+	compRegEx = regexp.MustCompile(`^(?P<bucket>.+)\.(?P<zone>.+)\.qingstor\.com$`)
+)
+
+
+func validateVisibility(visibility string) error {
+	switch visibility {
+	case "public":
+	case "private":
+	case "labeled":
```
Comment (zheng1) [inline pkg/manager/repo/validator.go]: What is the meaning of `labeled`?

File: pkg/manager/repo/validator.go [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+	"encoding/json"
+	"fmt"
+	"errors"
+	"strings"
+	"net/http"
+	"regexp"
+	neturl "net/url"
+	qs_conf "github.com/yunify/qingstor-sdk-go/config"
+	qs "github.com/yunify/qingstor-sdk-go/service"
+	"openpitrix.io/openpitrix/pkg/logger"
```
Comment (zheng1) [inline pkg/manager/repo/validator.go]: The code is not formatted. Please `git commit` after `make fmt-all`

File: vendor/vendor.json [hunk 6]
```diff
@@ -382,6 +382,18 @@
 			"revision": "7d79101e329e5a3adf994758c578dab82b90c017",
 			"revisionTime": "2016-05-24T15:18:35Z"
 		},
+		{
+			"checksumSHA1": "e0ISYl7mw9BaAUOidoOpdwIEGPQ=",
+			"path": "github.com/google/go-github/github",
```
Comment (zheng1) [inline vendor/vendor.json]: Why do we need this package?

File: pkg/manager/repo/validator.go [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+)
+
+
+var(
+	compRegEx = regexp.MustCompile(`^(?P<bucket>.+)\.(?P<zone>.+)\.qingstor\.com$`)
```
[the remark this answers is not in the document]
Comment (Hevienz) [reply pkg/manager/repo/validator.go]: Please give me a bucket url in private cloud.

File: pkg/manager/repo/validator.go (hunk 4 above)
Comment (Hevienz) [reply pkg/manager/repo/validator.go]: When a repo is labeled, user can access it if his repo-selector has a label match it.

File: vendor/vendor.json (hunk 6 above)
Comment (Hevienz) [reply vendor/vendor.json]: No longer used.

File: vendor/github.com/google/go-github/LICENSE [hunk 8]
```diff
@@ -0,0 +1,341 @@
+Copyright (c) 2013 The go-github AUTHORS. All rights reserved.
```
Comment (zheng1) [inline vendor/github.com/google/go-github/LICENSE]: Remove useless vendors

File: vendor/vendor.json [hunk 9]
```diff
@@ -587,6 +629,72 @@
 			"revision": "efb555f6a4e9e5841f162d37bd0684fbf9cc4984",
 			"revisionTime": "2017-12-28T16:00:23Z"
 		},
+		{
+			"checksumSHA1": "YKIcl6SFm1Fc9UT4fptu34eqEC4=",
+			"path": "github.com/yunify/qingstor-sdk-go",
```
Comment (chilianyi) [inline vendor/vendor.json]: In alpha version, we will support qingstor as repo address, but the protocol is s3.
In beta version, we will support other s3 protocol address, such as aws object storage. 
So in current alpha version, you need to visit qingstor with s3 sdk. These qingstor sdk should be deleted.

File: pkg/manager/repo/handler.go [hunk 10]
```diff
@@ -67,17 +84,27 @@ func (p *Server) DescribeRepos(ctx context.Context, req *pb.DescribeReposRequest
 }
 
 func (p *Server) CreateRepo(ctx context.Context, req *pb.CreateRepoRequest) (*pb.CreateRepoResponse, error) {
-	// TODO: validate CreateRepoRequest
+	typ := req.GetType().GetValue()
```
[the remark this answers is not in the document]
Comment (Hevienz) [reply pkg/manager/repo/handler.go]: When I use type as a variable name, compile will fail.

File: pkg/manager/repo/utils.go [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+		i++
+	}
+
+	return t
+}
+
+func GenerateLabelFilter(table, keyfield, valuefield string, labels_slice [][]interface{}) string {
+	var lf string
+
+	c := cartesian.Iter(labels_slice...)
```
[the remark this answers is not in the document]
Comment (Hevienz) [reply pkg/manager/repo/utils.go]: I found this is not work.

The generated SQL is "SELECT repo.repo_id, repo.name, repo.description, repo.type, repo.url, repo.credential, repo.visibility, repo.owner, repo.status, repo.create_time, repo.status_time, repo_label.label_key, repo_label.label_value FROM repo JOIN `repo_label` ON repo_label.repo_id = repo.repo_id WHERE ((repo_label.label_key = 'runtime' and repo_label.label_value in ('aws')) AND (repo_label.label_key = 'env' and repo_label.label_value in ('test'))) LIMIT 20 OFFSET 0;".

But a repo with label "runtime=aws" and "env=test" not returned as expect.

[the remark this answers is not in the document]
Comment (zheng1) [reply pkg/manager/repo/utils.go]: I had tested the following SQL, it is workable.

```sql
select repo.* from repo 
join repo_label as l1 on repo.repo_id = l1.repo_id 
join repo_label as l2 on repo.repo_id = l2.repo_id 
where (
(l1.label_key = "runtime" and l1.label_value in ("aws")) 
and 
(l2.label_key = "env" and l2.label_value in ("test"))
)
;
```

File: pkg/manager/repo/handler.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (zheng1) [reply pkg/manager/repo/handler.go]: You can change it to `repoType`

File: pkg/db/schema/repo/V0_1__init.sql [hunk 12]
```diff
@@ -2,6 +2,7 @@ CREATE TABLE repo (
 	repo_id     VARCHAR(50) PRIMARY KEY             NOT NULL,
 	name        VARCHAR(255)                        NOT NULL,
 	description TEXT                                NOT NULL,
+	type        VARCHAR(50)                         NOT NULL,
```
Comment (zheng1) [inline pkg/db/schema/repo/V0_1__init.sql]: Move this line to V0_2__add_repo_type.sql:
```sql
alter table repo add type VARCHAR(50) NOT NULL;
```

File: pkg/manager/repo/utils.go [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+	"net/url"
+	"strings"
+
+	"openpitrix.io/openpitrix/pkg/db"
+	"openpitrix.io/openpitrix/pkg/pb"
+)
+
+func GetLabelMapFromRequest(req *pb.DescribeReposRequest) (map[string][]string, error) {
+	lbs := req.GetLabel().GetValue()
+	lbs = strings.Replace(lbs, ",", "&", -1)
```
Comment (zheng1) [inline pkg/manager/repo/utils.go]: This line is not necessary, `url.ParseQuery` will decode labels

File: pkg/manager/repo/utils.go [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	return lbm, nil
+}
+
+func GetSelectorMapFromRequest(req *pb.DescribeReposRequest) (map[string][]string, error) {
+	sls := req.GetSelector().GetValue()
+	sls = strings.Replace(sls, ",", "&", -1)
```
Comment (zheng1) [inline pkg/manager/repo/utils.go]: ditto

**Pull request merged**